### PR TITLE
feat(tracks): add follow chain propagation and intent-based sampling [TRL-107]

### PR DESCRIPTION
**Pull request by galligan**
## Summary

- **TraceContext** with traceId, spanId, sampled flag — stored in `ctx.extensions[TRACE_CONTEXT_KEY]`
- **getTraceContext/childTraceContext** for parent-child inheritance through follow chains
- **Intent-based sampling**: read 5%, write 100%, destroy 100%. Decision at root, inherited by children.
- **Error promotion**: sampled-out traces promoted to sampled if any record errors (`keepOnError: true` default)
- Backward compatible: no sampling config = record everything

## Test plan

- [ ] 5 trace-context tests + 9 sampling tests + 3 layer integration tests
- [ ] `bun test` passes in `packages/tracks/`

Closes https://linear.app/outfitter/issue/TRL-107/follow-chain-propagation-and-intent-based-sampling

In-collaboration-with: [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by galligan** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/53?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#56** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/56?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#55** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/55?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#54** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/54?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#53** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/53?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/53?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#52** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/52?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#51** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/51?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#50** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/50?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#57](https://github.com/outfitter-dev/trails/pull/57) <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/57?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#43** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/43?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#44](https://github.com/outfitter-dev/trails/pull/44) <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/44?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#59** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/59?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#36](https://github.com/outfitter-dev/trails/pull/36) <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/36?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#35** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/35?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#34** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/34?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#33** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/33?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#32** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/32?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#31** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/31?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#30** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/30?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#29** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/29?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#28** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/28?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#27** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/27?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#26** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/26?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#25** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/25?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/24?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#23** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/23?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#22** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/22?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/21?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#19** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/19?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#20](https://github.com/outfitter-dev/trails/pull/20) <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/20?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#18** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/18?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#17** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/17?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#16** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/16?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#15** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/15?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#14** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/14?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#13** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/13?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#12** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/12?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#11** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/11?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#10** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/10?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#9** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/9?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/8?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/7?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/6?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#5** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/5?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/4?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/3?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/2?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1** <a href="https://app.graphite.com/github/pr/outfitter-dev/trails/1?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/tracks/src/tracks-layer.ts [hunk 1]
```diff
@@ -65,26 +144,45 @@ const buildEnrichedContext = (
...
+      const sampled = resolveSampled(
+        parentTrace,
+        trail.intent,
+        options?.sampling
+      );
+      const trace = resolveTrace(parentTrace, sampled);
+      const enriched = buildEnrichedContext(ctx, trace, sink);
       const record = createTrackRecord({
         intent: trail.intent,
+        parentId: parentTrace?.spanId,
```
[the remark this answers is not in the document]
Comment (galligan) [reply packages/tracks/src/tracks-layer.ts]: By design: `parentId` holds `parentTrace.spanId` from `TraceContext`, which is the identifier used consistently for trace correlation throughout the layer. Changing it would break the trace model.

File: packages/tracks/src/tracks-layer.ts [hunk 2]
```diff
@@ -35,6 +47,73 @@ const deriveOutcome = (result: Result<unknown, Error>): TrackOutcome =>
...
+/** Decide whether a completed record should be written to the sink. */
+const shouldWrite = (
+  record: TrackRecord,
+  sampled: boolean,
+  keepOnError: boolean
+): boolean => {
+  if (sampled) {
+    return true;
+  }
+  return keepOnError && record.status === 'err';
```
[the remark this answers is not in the document]
Comment (galligan) [reply packages/tracks/src/tracks-layer.ts:126]: Acknowledged. The `keepOnError` span-vs-trace promotion semantics are a design clarification, not a simple bug fix. Deferring.

**Comment by galligan** [comment]
Added the regression coverage for the C3 surface-wiring follow-up here. The tracks-layer test suite now exercises surface capture from context, so we have a direct guardrail around the new  plumbing as the follow-chain and sampling work on this branch evolves.

**Comment by galligan** [comment]
Added the regression coverage for the C3 surface-wiring follow-up here. The tracks-layer test suite now exercises surface capture from context, so we have a direct guardrail around the new `TrackRecord.surface` plumbing as the follow-chain and sampling work on this branch evolves.

**Comment by galligan** [comment]
Addressed the latest review feedback in the new push.

What changed:
- linked child records to parent record IDs instead of transient trace span IDs
- kept empty `sampling: {}` on the "record everything" path instead of silently activating default read sampling
- routed sink write failures through `onSinkError` while still preserving trail result delivery
- cleaned up the `prepareExecution` / `createTracksLayer` doc placement
- updated tests to cover parent linkage, surface capture, error promotion, and the empty-sampling edge case

Why:
- this branch is where follow-chain propagation and sampling semantics become user-visible, so the parentage and sampling behavior need to be correct here rather than deferred to later branches
- the older missing-module / parent-link findings should now be stale once Greptile reruns on the latest head

@greptileai review
@codex review
